### ISSUE TITLE
Lab 03 need some help with order Priority

### ISSUE THREAD

**Issue by aldensiol** [nus-cs2030/2223-s2#112]
Hi guys i'm stuck at implementing my logic.
The flow is as such:

Arrive -> Serve -> Done (this is ok)
or 
Arrive - > Leave (this is ok)
or
Arrive -> Wait (toString() will return “wait at”) -> PendingServe (returns “” empty string -- waits for previous serveEvent to be executed) -> Serve (current customer) -> Done

My serverList is only altered after my serveEvent is executed (which returns a doneEvent)

Since the pendingServeEvent is currently attached to a specific Server object, how should i make it such that the previous serveEvent (who is being served in the same Server object) first returns a doneEvent, and this pendingServeEvent goes straight to the front of the queue immediately after?

**Comment by aldensiol** [comment]
I should probably add that my server objects aren't instantiated by Customer objects, but by :
int serverID, double nextAvailTime, int numOfCustQueue

**Comment by granolaluvr69** [comment]
make the pendingServeEvent's timestamp the nextAvailTime of the server.

**Comment by aldensiol** [comment]
> make the pendingServeEvent's timestamp the nextAvailTime of the server.



> You can try to make the pendingServeEvent's timestamp the nextAvailTime of the server.

yep but since my waitEvent's execute() method returns a pendingServeEvent, at that moment in time i dont yet have the correct finalTime (time after the previousCust is served)

**Comment by zxciisdbniel** [comment]
> You can try to make the pendingServeEvent's timestamp the nextAvailTime of the server.

On top of this you probably need something to check that the server is going to serve this particular event next. (analogy like a queue number?) 
So your pendingServe event probably has to return itself with an updated timing (the new nextAvailTime) if the queue number doesn't match the server's yet.

**Comment by aldensiol** [comment]
hmm will definitely try out pendingServer returning itself with an updated timing

**Comment by aldensiol** [comment]
> > You can try to make the pendingServeEvent's timestamp the nextAvailTime of the server.
> 
> On top of this you probably need something to check that the server is going to serve this particular event next. (analogy like a queue number?) So your pendingServe event probably has to return itself with an updated timing (the new nextAvailTime) if the queue number doesn't match the server's yet.

appreciate it

**Comment by nicholasteow** [comment]
hi alden u could perhaps try assign additional values that is specific to the nature of the event to the particular event to ensure the order of events to be printed smoothly.

Do take note to be extremely careful with these values as they should not be able to change the original value and should work for any input values not just those provided by test cases.

**Comment by aldensiol** [comment]
> 

thank you nicholas, will try it out!

**Comment by nicholasteow** [comment]
Welcome 👯‍♂️

**Comment by granolaluvr69** [comment]
> > make the pendingServeEvent's timestamp the nextAvailTime of the server.
> 
> > You can try to make the pendingServeEvent's timestamp the nextAvailTime of the server.
> 
> yep but since my waitEvent's execute() method returns a pendingServeEvent, at that moment in time i dont yet have the correct finalTime (time after the previousCust is served)

In this case, you can try updating serverlist every for every iteration of your loop in simulate(), and obtain the server's updated attributes for every event from that updated serverlist when you need to create a new event.

**Comment by aldensiol** [comment]
> > > make the pendingServeEvent's timestamp the nextAvailTime of the server.
> > 
> > 
> > > You can try to make the pendingServeEvent's timestamp the nextAvailTime of the server.
> > 
> > 
> > yep but since my waitEvent's execute() method returns a pendingServeEvent, at that moment in time i dont yet have the correct finalTime (time after the previousCust is served)
> 
> In this case, you can try updating serverlist every for every iteration of your loop in simulate(), and obtain the server's updated attributes for every event from that updated serverlist when you need to create a new event.

Hi kenneth, thank you so much for the tip. I can tell that you are quite the handsome guy.

**Comment by haqanh** [comment]
> Hi guys i'm stuck at implementing my logic. The flow is as such:
> 
> Arrive -> Serve -> Done (this is ok) or Arrive - > Leave (this is ok) or Arrive -> Wait (toString() will return “wait at”) -> PendingServe (returns “” empty string -- waits for previous serveEvent to be executed) -> Serve (current customer) -> Done
> 
> My serverList is only altered after my serveEvent is executed (which returns a doneEvent)
> 
> Since the pendingServeEvent is currently attached to a specific Server object, how should i make it such that the previous serveEvent (who is being served in the same Server object) first returns a doneEvent, and this pendingServeEvent goes straight to the front of the queue immediately after?

would it be possible for you to obtain the done time of your server (since you are returning the Event from here) and then make this the timestamp of the Wait event you are returning?

**Comment by aldensiol** [comment]
> > Hi guys i'm stuck at implementing my logic. The flow is as such:
> > Arrive -> Serve -> Done (this is ok) or Arrive - > Leave (this is ok) or Arrive -> Wait (toString() will return “wait at”) -> PendingServe (returns “” empty string -- waits for previous serveEvent to be executed) -> Serve (current customer) -> Done
> > My serverList is only altered after my serveEvent is executed (which returns a doneEvent)
> > Since the pendingServeEvent is currently attached to a specific Server object, how should i make it such that the previous serveEvent (who is being served in the same Server object) first returns a doneEvent, and this pendingServeEvent goes straight to the front of the queue immediately after?
> 
> would it be possible for you to obtain the done time of your server (since you are returning the Event from here) and then make this the timestamp of the Wait event you are returning?

hmm well since my WaitEvent's toString() should include initial waiting time, i don't think it would be correct to do so.
